Make frame boxes no longer TileEntities

NotAPenguin0 opened the pull request:
This PR makes frame boxes regular blocks instead of TileEntities. They still accept covers like before, by spawning a TileEntity the moment they are right clicked with a cover. The majority of frame boxes in real bases won't need these of course, so this should improve lag significantly. Some notes:

- ~~The TileEntity version of the frame box item still exists, but is hidden in NEI, no longer craftable and no longer used in recipes. This is a pretty big breaking change, since players will need to update all their AE patterns dealing with frames. If anyone knows a way around this, please let me know~~. Postea ItemStack transformer fully converts all frame box items, even those in AE patterns.
- ~~There's currently an issue where frame boxes seem to forget they have covers attached on world load. The tileentity still exists, because I can right click the frame with a crowbar and the cover comes off, or I can add a new cover and the old ones show up as well. Might just be a rendering issue, but if anyone knows how to fix this I'd be grateful.~~ Rendering issue that will be fixed soon.
- On drop, frame boxes no longer keep all their cover settings, but instead simply drop the cover items they had attached to them. I think this is a fine tradeoff honestly.
- I couldn't get the correct breaking tool and harvest level to work, so someone please look over that.
- ~~Existing frameboxes will need to be converted to the block version, @Connor-Colenso has told me to do this using postea, so I'll try that in a different PR over there I think? Or is this code that should go in GT as well?~~ TileEntity transformer works.
- ~~Probably also need to remove the glass oredict, but I haven't figured out how yet.~~

![image](https://github.com/user-attachments/assets/8e005f07-ec89-45b2-a1e4-35d83b2299fa)
![image](https://github.com/user-attachments/assets/c62f386b-ff25-453c-9e34-eb2e335746db)

Dream-Master commented:
draft until the structure check work

OneEyeMaker commented on src/main/java/gregtech/api/util/GT_StructureUtility.java:
```diff
@@ -74,11 +73,11 @@ public static <T> IStructureElement<T> ofFrame(Materials aFrameMaterial) {
-                    if (tPipeBase.isInvalidTileEntity()) return false;
-                    if (tPipeBase.getMetaTileEntity() instanceof GT_MetaPipeEntity_Frame)
-                        return aFrameMaterial == ((GT_MetaPipeEntity_Frame) tPipeBase.getMetaTileEntity()).mMaterial;
+                Block block = world.getBlock(x, y, z);
+                if (block instanceof GT_Block_FrameBox frameBox) {
+                    int meta = world.getBlockMetadata(x, y, z);
+                    Materials material = frameBox.getMaterial(meta);
+                    return aFrameMaterial == material;
```
I think this change should be backward compatible (to prevent sudden breakage of existing multiblocks).

NotAPenguin0 replied:
I plan on writing a transformer using https://github.com/gtNewHorizons/postea that turns all in world TE frameboxes into their new block versions, which should solve this

NotAPenguin0 commented:
Aside from one rendering issue where covers are not rendered after reloading the world until the framebox is modified in some way, this should be ready now.

NotAPenguin0 commented:
Seems like I need to fix a crash somewhere first

NotAPenguin0 commented:
The difference is that as far as I know, ores always need that attached TileEntity to function properly, and I don't know enough about their behaviour to change that. The reason this whole thing can be removed from frame boxes is because they only need their TileEntity in rare cases and not by default, when they need this TE it's still spawned even with this PR.

mitchej123 commented:
> Hi! can you make ores be also regular block? because ores are also tile entities, i know it might be useless question/proposal, but it makes impossible to use cofh world or condensedores to make custom oregen with Gregtech

This doesn't belong here on an unrelated PR.

It needs a separate issue/PR, and might already have one.

NotAPenguin0 commented:
Yeah if anything it should go in a different PR

mitchej123 commented:
> but i can make issue, in case someone interested in remaking ores to be regular blocks

Check that there isn't one already.  Colen has wanted to do it and postea was implemented for that reason.  I know it hasn't yet been pushed forward on the implementation side.

NotAPenguin0 commented:
This also modifies the tesla tower code slightly, for two reasons:
- It used a custom implementation of `StructureUtility.ofFrame()`, making it unable to form with the new frame boxes. Simply switching to `ofFrame()` works.
- It crashed the machine on shutdown because of a missing null check so I just added that fix in here.

OneEyeMaker commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -0,0 +1,455 @@
+            default -> ForgeDirection.UNKNOWN;
+        };
+        final TileEntity frameEntity = aWorld.getTileEntity(aX, aY, aZ);
+        if (frameEntity == null) return false;
```
This line is not necessary since `instanceof` check below will do the trick.

OneEyeMaker commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -0,0 +1,455 @@
+    @Override
+    public void onEntityCollidedWithBlock(World aWorld, int aX, int aY, int aZ, Entity collider) {
+        final TileEntity tTileEntity = aWorld.getTileEntity(aX, aY, aZ);
+        if (tTileEntity instanceof IGregTechTileEntity
+            && ((IGregTechTileEntity) tTileEntity).getMetaTileEntity() != null) {
+            ((IGregTechTileEntity) tTileEntity).onEntityCollidedWithBlock(aWorld, aX, aY, aZ, collider);
```
```suggestion
        if (tTileEntity instanceof IGregTechTileEntity gtTE
            && gtTE.getMetaTileEntity() != null) {
            gtTE.onEntityCollidedWithBlock(aWorld, aX, aY, aZ, collider);
```

OneEyeMaker replied:
Same for other methods.

OneEyeMaker commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -0,0 +1,455 @@
+            return true;
+        }
+        final TileEntity tTileEntity = aWorld.getTileEntity(aX, aY, aZ);
+        if (tTileEntity != null) {
```
Null-check is not necessary.

OneEyeMaker commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -0,0 +1,455 @@
+
+    @Override
+    public IIcon getIcon(int side, int meta) {
+        // This doesn't apply the frame color so AE2 facades look bad, but it renders fine in inventory and in world.
+        // It's probably fine, who makes facades out of frame boxes anyway.
```
IIRC, `getRenderColor(int meta)` is responsive for that. Try override this method and retrieve color from material.

NotAPenguin0 replied:
I've tried implementing this, but it still doesn't seem to work very well and leaves them all white.

```java
    @Override
    public int getRenderColor(int meta) {
        Materials material = getMaterial(meta);
        short[] rgba = material.mRGBa;
        return (rgba[0] << 24) + (rgba[1] << 16) + (rgba[2] << 8) + rgba[3];
    }
```

I'm not 100% sure this implementation is correct, though if it ends up not working I think it's somewhat fine since these cover items are always hidden in NEI and I can't imagine anyone using them. Would be nice to get it fixed either way though, so if you can spot what's wrong here please let me know.

OneEyeMaker commented on src/main/java/gregtech/loaders/preload/GT_Loader_MetaTileEntities.java:
```diff
@@ -11886,19 +11894,43 @@ private static void registerPlasmaGenerators() {
-                || (GregTech_API.sGeneratedMaterials[i] == Materials.Wood)) {
+        for (int meta = 0; meta < GregTech_API.sGeneratedMaterials.length; meta++) {
+            Materials material = GregTech_API.sGeneratedMaterials[meta];
+            if (((GregTech_API.sGeneratedMaterials[meta] != null)
+                && ((GregTech_API.sGeneratedMaterials[meta].mTypes & 0x2) != 0))
+                || (GregTech_API.sGeneratedMaterials[meta] == Materials.Wood)) {
```
```suggestion
            if ((material != null && (material.mTypes & 0x2) != 0)
                || material == Materials.Wood) {
```

OneEyeMaker replied:
And keep using this local variable below.

OneEyeMaker reviewed (changes requested):
Looks good overall. Just needs a code cleanup.

OneEyeMaker commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -0,0 +1,455 @@
+    private static final ThreadLocal<IGregTechTileEntity> mTemporaryTileEntity = new ThreadLocal<>();
+
+    public GT_Block_FrameBox() {
+        super(Material.glass);
```
Please, change material of block. Otherwise BW will mark this block as HV glass.

OneEyeMaker replied:
Original material was `GT_Material_Machines`.

OneEyeMaker commented:
There are 2 minor visual inconsistencies:
* Because new frames use a different `ItemBlock` class, the original tooltip is missing (`MetaPipeEntity_Frames::getDescription`).
* Frames without a TileEntity don't display a grid when hovering over the block with a tool or cover.

NotAPenguin0 commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -0,0 +1,455 @@
+
+    @Override
+    public int getHarvestLevel(int aMeta) {
+        return aMeta % 4;
```
I wasn't sure what to do for this, I think I copied the implementation from some other block, but I agree `aMeta % 4` probably doesn't make much sense, since `aMeta` is just the material index. Maybe there's a way to get a harvest level from the material, but if not I'd probably just make this a single fixed harvest level.

OneEyeMaker replied:
AFAIK `Material::mToolQuality` describes the mining level of a *tool*, not the block made from the material. I think keeping it as it is now would be fine because one is equivalent to the old value.

NotAPenguin0 commented:
![image](https://github.com/user-attachments/assets/5e4dae02-c892-4e52-a4cd-dbb0cbcefd32)
I've added the description and removed the glass material. Also, the wrench is now properly required to break a frame box, but the harvest level does still need work. 

I'm not sure what to do about displaying the grid really, I agree it's a bit of a weird inconsistency so it would be nice to have, but I'm not sure how without digging more into the arcana that is GT rendering code.

Ethryan commented:
Psst @NotAPenguin0 
https://github.com/GTNewHorizons/GT5-Unofficial/blob/master/src/main/java/gregtech/common/GT_Client.java#L309
Got something for you.

Caedis commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -0,0 +1,442 @@
+    @Override
+    public float getPlayerRelativeBlockHardness(EntityPlayer aPlayer, World aWorld, int aX, int aY, int aZ) {
+        final TileEntity tTileEntity = aWorld.getTileEntity(aX, aY, aZ);
+        return tTileEntity instanceof BaseMetaTileEntity && ((BaseMetaTileEntity) tTileEntity).privateAccess()
+            && !((BaseMetaTileEntity) tTileEntity).playerOwnsThis(aPlayer, true) ? -1.0F
```
instanceof casts is probably a better name to describe what you are wanting.

OneEyeMaker commented:
![2024-08-24_09 17 06](https://github.com/user-attachments/assets/4779e4c1-2798-4d76-9259-3c9ffdc921d0)

Add this method to `GT_Item_Frames`:
```java
    @Override
    public int getColorFromItemStack(ItemStack aStack, int aPass) {
        int meta = aStack.getItemDamage();
        Materials material = ((GT_Block_FrameBox) this.block()).getMaterial(meta);
        return (material.mRGBa[0] << 16) | (material.mRGBa[1] << 8) | material.mRGBa[2];
    }
```

Method `getRenderColor` can be safely removed from `GT_Block_FrameBox`.
Also, what about making `getMaterial` method static (to prevent such ugly casts)?

NotAPenguin0 commented:
![image](https://github.com/user-attachments/assets/31e37b39-0382-4d45-aa85-5c398bbfb299)
Much better, thanks @Ethryan

OneEyeMaker commented on src/main/java/gregtech/common/GT_Client.java:
```diff
@@ -847,6 +848,14 @@ public void onDrawBlockHighlight(DrawBlockHighlightEvent aEvent) {
         }
 
+        // If there is no tile entity and the block is a frame box block, still draw the grid if a cover is held
+        if (aTileEntity == null && aBlock instanceof GT_Block_FrameBox) {
```
I suggest adding a grid when hovering over the block with the wrench. Just add this condition to the `if` expression above.

NotAPenguin0 replied:
I thought about this, but since you can't really rotate the block frame boxes I'm not sure if this makes a lot of sense to do

NotAPenguin0 replied:
![image](https://github.com/user-attachments/assets/36b4dfae-404b-4439-a215-e40661af2ce3)
Adding it was easy though and I guess it doesn't hurt to have

OneEyeMaker replied:
This grid should tell players that the wrench is the correct tool to *break* the block. Also, the grid should be present on regular blocks for compatibility with blocks with TileEntity. 
Grid is always displayed on TileEntity (regardless of the existence of covers). So it should be on a regular block too.

OneEyeMaker commented on src/main/java/gregtech/common/blocks/GT_Item_Frames.java:
```diff
@@ -49,7 +49,7 @@ public int getColorFromItemStack(ItemStack aStack, int aPass) {
     @Override
     public void addInformation(ItemStack aStack, EntityPlayer aPlayer, List<String> aList, boolean aF3_H) {
-        aList.add(GT_LanguageManager.getTranslation("gt.blockmachines.gt_frame.desc.format"));
+        aList.add(GT_LanguageManager.getTranslation("gt.blockframes." + aStack.getItemDamage() + ".tooltip"));
```
Why do you replace description when I suggest to just extend it? ;) IMHO it would be nice to have both lines (chemical element localized with key `gt.blockframes.<meta>.tooltip` and info about covers - `gt.blockmachines.gt_frame.desc.format`).

Anyway I leave this up to you.

NotAPenguin0 replied:
Oh yeah it could totally have both actually, I forgot I can add multiple lines

NotAPenguin0 replied:
![image](https://github.com/user-attachments/assets/196a758b-0ce4-4ab3-833d-274d4299bc0c)
That's better

OneEyeMaker reviewed (approved):
Great job!

NotAPenguin0 commented:
Adding one more fix to make breaking and placing frame boxes cause structure updates like regular casing blocks do

OneEyeMaker commented on src/main/java/gregtech/common/blocks/GT_Block_FrameBox.java:
```diff
@@ -214,9 +214,18 @@ public void breakBlock(World aWorld, int aX, int aY, int aZ, Block aBlock, int a
             mTemporaryTileEntity.set(gtTE);
         }
+        // Cause structure update
+        GregTech_API.causeMachineUpdate(aWorld, aX, aY, aZ);
```
Should be called before `onBlockDestroyed` (according to `GT_Block_Machines`)
